src/main/java/de/unistuttgart/iste/gits/content_service/service/SuggestionService.java

```diff
@@ -0,0 +1,145 @@
+package de.unistuttgart.iste.gits.content_service.service;
+
+import de.unistuttgart.iste.gits.generated.dto.*;
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.lang3.tuple.Pair;
+import org.springframework.stereotype.Service;
+
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.util.*;
+import java.util.stream.Stream;
+
+import static java.time.OffsetDateTime.now;
+import static java.util.Comparator.comparing;
+import static java.util.Comparator.reverseOrder;
+import static org.apache.commons.lang3.ObjectUtils.isEmpty;
+
+/**
+ * Service for creating suggestions for the user.
+ */
+@Service
+@RequiredArgsConstructor
+public class SuggestionService {
+
+    private final SectionService sectionService;
+    private final UserProgressDataService userProgressDataService;
+
+    private final Map<Pair<UUID, UUID>, UserProgressData> userProgressDataCache = new HashMap<>();
+
+    /**
+     * Creates {@link Suggestion}s for the given chapter IDs and user ID. The suggestions are created based on the
+     * user's progress and the given skill types.
+     *
+     * @param chapterIds the IDs of the chapters for which suggestions should be created.
+     * @param userId     the ID of the user for which suggestions should be created.
+     * @param amount     the amount of suggestions to create.
+     * @param skillTypes the skill types of the suggestions to create. If the list is empty, all skill types are
+     *                   considered. If the list is not empty, only the given skill types are considered and
+     *                   media contents are ignored.
+     * @return the created suggestions.
+     */
+    public List<Suggestion> createSuggestions(List<UUID> chapterIds, UUID userId, int amount, List<SkillType> skillTypes) {
+        userProgressDataCache.clear();
+
+        List<Stage> availableStages = sectionService.getSectionsByChapterIds(chapterIds)
+                .stream()
+                .flatMap(Collection::stream)
+                .flatMap(section -> getAvailableStagesOfSection(section, userId).stream())
+                .toList();
+
+        return Stream.concat(
```

Comment:
Doesn't this mean that there exists a case where only optional content is suggested, even if there is required content ready for repetition?

Reply:
So you think, optional content should be prioritized lower or something? Good point, would have been nice to get this feedback in the concept pull request and not after implementing it 😅

I guess I need to rewrite most of it now...

Reply:
I think optional content could be suggested if there is still room left in the return list. 
So, in other words, scan required content first. Then, if there is still room, fill up the remaining spots left in the return list with optional content that is due. 
What do you think of that approach?

Reply:
fixed it like you said, makes sense! :)

Reply:
If I understand your new implementation correctly, you now always take all required content, sort & filter them by skilltypes and then repeat the same and put the optional content after the required content.
But doesn't this mean, we don't check if the content is even due for repetition yet?

Also, this means optional content would never be reachable if sample size is smaller than required content list size

Reply:
```
.sorted(comparing((Content content) -> Duration.between(now(), getRelevantLearnDate(content, userId)).toDays())
```
line 81 or so, there is your check for the due date

Reply:
> Also, this means optional content would never be reachable if sample size is smaller than required content list size

I mean okay I could make an performance improvement that the optional part is only evaluated if necessary, but that is done by the java stream api automatically, as far as I know.
The optional contents wont be filtered and sorted, if the required contents are enough